### PR TITLE
Feed forward hotend regulator (so far)

### PR DESCRIPTION
**Pull request by thinkyhead**
Rebased #16953 onto `bugfix-2.0.x`.

---

### Requirements

Do not merge yet. It doesn't fulfill Marlin coding style. It can not be enabled/disabled by macro. It contains hard coded constants for single printer model. It is in conflict / duplicate to `PID_FAN_SCALING`. Probably wont compile against recent Marlin. It is probably not opened against the right branch. You can see working example there https://github.com/prusa3d/Prusa-Firmware-Buddy/pull/154

### Description

Implements feed forward regulator based on print head thermal capacity and cooling model. PID regulator is used only for feedback fine tuning of output power. This allows adding additional safety check, which compares expected vs. actual power output. Such check can detect very rare but already observed situation, where thermistor resistance locks very near to target temperature leading to uncontained thermal runaway.

If you express interest in merging some of the features (eg. model based feed forward regulator, additional safety check(not implemented yet), different I therm limitation, different D therm computation) I am willing to rewrite such features to match Marlin coding standards, ability to been disabled by macro. Move hardcoded constants to been possibly configured via preprocessor macro or EEPROM storage. And write manual how to tune such constant or write autotune routine.

It is currently undergoing testing at Prusa test department. Once I have results I will add similar temperature charts for new implementation.
This is state before. It is not zipped, just rename to MINI_PET_old_implementation_english.ods
[MINI_PET_old_implementation_english.ods.zip](https://github.com/MarlinFirmware/Marlin/files/4247505/MINI_PET_old_implementation_english.ods.zip)

### Benefits

 - Improved safety with single temperature sensor (planned, not implemented yet)
 - Improved temperature regulation with less over/under shots.
 - Save RAM consumption by omitting delay line for extrusion scaling.

### CODE REVIEW

**Comment by mkbel** [comment]
Does it mean, that you are interested in my proposal and I shoud rewrite the feature to be able to integrate it with current Marlin?

**Comment by mkbel** [comment]
Does it mean, that you are interested in my proposal and I shoud rewrite the feature to be able to integrate it with current Marlin?

**Comment by thinkyhead** [comment]
This is the third attempt to link material volume and other factors to heater power. Will it be better than the first two approaches? Only very rigorous testing of the different approaches can tell. And will it be adopted if it can't show demonstrable value? I think it will help if you can show the difference in the temperature graph with and without the feature, and present other data that shows how well it works and why it requires little or no tuning compared to other approaches….

**Comment by thinkyhead** [comment]
You will note that I've done most of the rewriting already. Review the current set of changes under the FILES tab above.

**Comment by AnHardt** [comment]
Would you please be so kind to explain:
* the general algorithm in words,
* the physics and math behind the model of  `ff_steady_state_hotend(float target_temp, float print_fan)`,
* how to get the constants.

**Comment by mkbel** [comment]
- Temperature graph of short PLA print on Prusa MINI before
image
image
image
- Temperature graph of short PLA print on Prusa MINI after
image
image
image

[MINI_PLA_before_en.ods.zip](https://github.com/MarlinFirmware/Marlin/files/4269387/MINI_PLA_before_en.ods.zip) LibreOffice Calc document, just rename to MINI_PLA_before_en.ods
[MINI_PLA_after_en.ods.zip](https://github.com/MarlinFirmware/Marlin/files/4269388/MINI_PLA_after_en.ods.zip) LibreOffice Calc document, just rename to MINI_PLA_after_en.ods

**Comment by mkbel** [comment]
I am unable to push commits into this branch, nor edit pull request description. So I will reopen my original pull request once I will have time to finish it.

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -202,13 +202,7 @@
 #endif
 
 #if ENABLED(PIDTEMP)
-  // Add an experimental additional term to the heater power, proportional to the extrusion speed.
-  // A well-chosen Kc value should add just enough power to melt the increased material volume.
-  //#define PID_EXTRUSION_SCALING
-  #if ENABLED(PID_EXTRUSION_SCALING)
-    #define DEFAULT_Kc (100) //heating power=Kc*(e_speed)
-    #define LPQ_MAX_LEN 50
-  #endif
```
Comment (mkbel) [inline Marlin/Configuration_adv.h:211]: Do not remove PID_EXTRUSION_SCALING.
I just suggested removing  LPQ_MAX_LEN feature. I can't imagine scenario where you want to lag additional heating after extruding has taken place. To me it looks like some remain of workaround where it was not possible to rig directly to stepper.position(E_AXIS) and it was rigged somewhere before planer queue.

And I changed meaning of DEFAULT_Kc.

    #if ENABLED(PID_EXTRUSION_SCALING)
        /**
         * Increase in PWM duty cycle needed to to extrude 1 mm^2 per second
         * of filament if extruder temperature is 1 Kelvin above ambient
         * temperature [s*K^-1*mm^-2]
         */
        #define DEFAULT_Kc 0.009517f
    #endif

File: Marlin/src/feature/joystick.cpp [hunk 2]
```diff
@@ -173,7 +173,7 @@ Joystick joystick;
 
     if (!UNEAR_ZERO(hypot2)) {
       current_position += move_dist;
-      const float length = sqrt(hypot2);
+      const float length = SQRT(hypot2);
```
Comment (mkbel) [inline Marlin/src/feature/joystick.cpp:176]: This looks completely unrelated to temperature regulator.

File: Marlin/src/gcode/config/M301.cpp [hunk 3]
```diff
@@ -34,11 +34,6 @@
  *   I[float] Ki term (unscaled)
  *   D[float] Kd term (unscaled)
  *
- * With PID_EXTRUSION_SCALING:
- *
- *   C[float] Kc term
```
Comment (mkbel) [inline Marlin/src/gcode/config/M301.cpp:39]: Do not remove.

File: Marlin/src/gcode/config/M301.cpp [hunk 4]
```diff
@@ -53,9 +48,6 @@ void GcodeSuite::M301() {
     if (parser.seen('P')) PID_PARAM(Kp, e) = parser.value_float();
     if (parser.seen('I')) PID_PARAM(Ki, e) = scalePID_i(parser.value_float());
     if (parser.seen('D')) PID_PARAM(Kd, e) = scalePID_d(parser.value_float());
-    #if ENABLED(PID_EXTRUSION_SCALING)
-      if (parser.seen('C')) PID_PARAM(Kc, e) = parser.value_float();
-    #endif
```
Comment (mkbel) [inline Marlin/src/gcode/config/M301.cpp]: Do not remove.

File: Marlin/src/gcode/config/M301.cpp [hunk 5]
```diff
@@ -69,9 +61,6 @@ void GcodeSuite::M301() {
     SERIAL_ECHOPAIR(" p:", PID_PARAM(Kp, e),
                     " i:", unscalePID_i(PID_PARAM(Ki, e)),
                     " d:", unscalePID_d(PID_PARAM(Kd, e)));
-    #if ENABLED(PID_EXTRUSION_SCALING)
-      SERIAL_ECHOPAIR(" c:", PID_PARAM(Kc, e));
-    #endif
```
Comment (mkbel) [inline Marlin/src/gcode/config/M301.cpp:77]: Do not remove.

File: Marlin/src/inc/Conditionals_adv.h [hunk 6]
```diff
@@ -39,7 +39,7 @@
   #undef FWRETRACT
   #undef PIDTEMP
   #undef AUTOTEMP
-  #undef PID_EXTRUSION_SCALING
```
Comment (mkbel) [inline Marlin/src/inc/Conditionals_adv.h:42]: Do not remove.

File: Marlin/src/module/configuration_store.cpp [hunk 7]
```diff
@@ -2604,10 +2586,6 @@ void MarlinSettings::reset() {
       PID_PARAM(Kp, e) = float(DEFAULT_Kp);
       PID_PARAM(Ki, e) = scalePID_i(DEFAULT_Ki);
       PID_PARAM(Kd, e) = scalePID_d(DEFAULT_Kd);
-      #if ENABLED(PID_EXTRUSION_SCALING)
-        PID_PARAM(Kc, e) = DEFAULT_Kc;
-      #endif
-
```
Comment (mkbel) [inline Marlin/src/module/configuration_store.cpp:2618]: Do not remove.

**Comment by mkbel** [comment]
@thinkyhead I think you can close this pull request. As I did lot of unrelated changes and it makes it difficult to merge, review and test. I will open several smaller pull requests. I did several small changes, which I consider very beneficial. And the big change (feed forward regulator) is not yet complete (as it is hardcoded for single printer model) and (especially the biggest planed improvement of safety check is not even started).

**Comment by mkbel** [comment]
@thinkyhead Thank you for your very fast response to my opened pull request and your proposed changes. I already backported some of them to my fork.

**Comment by mkbel** [comment]
By the way does anybody know, what is macro MIN_BED_POWER supposed to do? From its name and from code it looks like bed power is supposed to be no less than MIN_BED_POWER, but final constrain of pid_output looks like:
`pid_output = constrain(work_pid.Kp + work_pid.Ki + work_pid.Kd + float(MIN_BED_POWER), 0, MAX_BED_POWER);`
Where output is not constrained to MIN_BED_POWER but to 0.

**Comment by thinkyhead** [comment]
> I am unable to push commits into this branch

For future reference, you would check out my branch and then submit a pull request to it. When the PR is accepted its commits are added to the branch, and are thus also added to this PR.

> Where output is not constrained to `MIN_BED_POWER` but to 0.

`MIN_BED_POWER` is added to the PID output and it is the final sum of the components that is not allowed to be < 0.

> I already backported some of them to my fork.

Great! If you have any other corrections to send this way, don't hesitate.
